lib/lineup/src/lib.rs

```diff
@@ -26,5 +26,21 @@ pub mod upcalls;
 /// Type to represent a core id for the scheduler.
 type CoreId = usize;
 
+/// A utility function that converts a core_id (which may not be sequential)
+/// to a form more easily used for indexing
+#[inline(always)]
+pub fn core_id_to_index(core_id: CoreId) -> usize {
+    let machine_id = kpi::system::mid_from_gtid(core_id);
+    if machine_id == 0 {
+        // This is controller (should never happen) or non-rackscale
+        core_id
+    } else {
+        // The controller (which will never need this) is always machine_id 0, so
+        // decrement here for rackscale clients.
+        kpi::process::MAX_CORES_PER_MACHINE * (machine_id - 1)
+            + kpi::system::mtid_from_gtid(core_id)
```

Comment:
isn't this `machine_id`?

Reply:
Instead of `machine_id - 1`? If that's the question, it could be. However, as the controller (id 0) doesn't ever use userspace resources, we safely shift the machine_ids of the clients down by one.

Reply:
I just meant the `kpi::system::mtid_from_gtid(core_id)` expression, you assigned it as `machine_id` a bit above. So you can use that variable too if you want but not necessary it's not very computation heavy anyways :)